### DISCUSSION TITLE
Performing igl::copyleft::cgal::mesh_boolean() in a loop

### DISCUSSION THREAD

**Discussion by drathin4** [libigl/libigl#1806] [Q&A]
I'm working on a mesh boolean project and trying to implement boolean calls on all files present in the folder. This boolean can be either of type union or subtract. Since the number of files present and filenames is not known beforehand, I have to perform this in a `for loop`. And I'm using the call `igl::copyleft::cgal::mesh_boolean()`. So initially I unite body1 & body2, say the result is body12. Then in the loop, I unite body12 with body3 and so on.... Here the operation goes wrong. Uniting body12 with body3 gives errors - `Odd cell cycle detected!`. As such body3 doesn't have any issues. If I unite body2 & body3, the result body23 is correct. The only problem is uniting a new body with an existing body on whom a mesh boolean operation is already performed.

In the LibIGL tutorials, I saw usage of `igl::copyleft::cgal::CSGTree`. But that approach won't work in my case since I need to do it programmatically in a loop and not hard code the values. What is the solution to this ?

**Comment by alecjacobson** [comment]
Can you post a minimal example where you get that error?

**Comment by drathin4** [comment]
I'm attaching 2 STL files in this message. 1st is is the resultant body of unite operation between 0.stl & 1.stl named as Unite_0_1.stl. The 2nd is 2.stl. If we apply `igl::copyleft::cgal::mesh_boolean()` with unite operation, it results in the `Odd cell cycle detected!` issue when run in debug mode. This is happening for many other files as well.

[UniteExample.zip](https://github.com/libigl/libigl/files/6466475/UniteExample.zip)

**Reply by drathin4** [reply]
Hi, just wanted to post an update. 

I downloaded and installed MeshLab today and used a number of repairing/cleaning filters like `removal of duplicate faces/vertices, repairing non-manifold edges by removing faces/splitting vertices, repairing non-manifold vertices by splitting, removing T-Vertices by edge collapse/flip, removal of sliver faces etc.` After this operation I exported the Unite_0_1.stl file and did boolean unite on 2.stl using `igl::copyleft::cgal::mesh_boolean()` and it worked correctly.

I found that LibIGL does provide `igl::decimate() & igl::collapse_edge()` APIs. Do you think I should post-process every booleaned mesh while running in a loop, so that it works correctly ? I believe that after the initial mesh boolean operation, there are some defects in the resultant mesh that need to be repaired.

**Reply by drathin4** [reply]
2nd update:

I was able to perform the boolean operation correctly  between 0.stl, 1.stl & 2.stl. I basically used `igl::remove_duplicate_vertices() & igl::resolve_duplicated_faces()` as a post-processing step after performing `igl::copyleft::cgal::mesh_boolean()` because there are duplicate vertices and faces after the boolean operation.

Now I have a bunch of files whose mesh data needs to united together. Among those files all the unite operations are working correctly barring one. I'm trying to figure out the reason for that. I have a list of 40 STL files and mesh data of 39 files is getting united properly. There's just 1 file other file which cannot be united to the aggregate union of 39 files. I'm currently running it in debug mode. In this case, even `igl::remove_duplicate_vertices() & igl::resolve_duplicated_faces()` is not working.

**Reply by alecjacobson** [reply]
Can you upload 0.stl and 1.stl separately?

**Reply by drathin4** [reply]
Attached herewith are 0.stl &1.stl in the .zip file

[0_1.zip](https://github.com/libigl/libigl/files/6498913/0_1.zip)

**Comment by drathin4** [comment]
@alecjacobson 

Finally, I have zeroed in on the problematic file - 14.stl. The 1st boolean unite operation itself (`igl::copyleft::cgal::mesh_boolean()`) doesn't happen correctly on the files 0.stl & 14.stl and they end up with the error: `Odd cell cycle detected!` when run in debug mode. I tried using the APIs `igl::is_irregular_vertex() & igl::is_vertex_manifold()`, but the file 14.stl didn't detect any such problems.

Could you please look into this and tell me if 14.stl is a defective file ? If so, can any mesh repair APIs be used to correct it ? I'm attaching the files: 0.stl & 14.stl with this message.

[ProblemFiles.zip](https://github.com/libigl/libigl/files/6487054/ProblemFiles.zip)

**Reply by alecjacobson** [reply]
In this example, `0.stl` is not [pwn](http://www.cs.columbia.edu/cg/mesh-arrangements/#definitions). This is a requirement for inputs to `igl::copyleft::cgal::mesh_boolean()`. 

Some mess of conflictingly oriented edges, highlighted here in orange:
![image](https://user-images.githubusercontent.com/2241689/118568302-8e1fb080-b745-11eb-81ce-05f26a6a6927.png)

**Reply by drathin4** [reply]
@alecjacobson 

Can we use any of the LibIGL APIs as a pre-processing step to repair the conflicting edges in the mesh so that it becomes PWN ?
